linker/shared/src/test/scala/org/scalajs/linker/IRCheckerTest.scala

```diff
@@ -171,13 +174,12 @@ object IRCheckerTest {
       .withCheckIR(true)
       .withOptimizer(false)
     val linkerFrontend = StandardLinkerFrontend(config)
-    val symbolRequirements = StandardLinkerBackend(config).symbolRequirements
 
     val classDefsFiles = classDefs.map(MemClassDefIRFile(_))
 
     val result = TestIRRepo.minilib.stdlibIRFiles.flatMap { stdLibFiles =>
       linkerFrontend.link(stdLibFiles ++ classDefsFiles, moduleInitializers,
-        symbolRequirements, new ScalaConsoleLogger(Level.Error))
+        noSymbolRequirements, new ScalaConsoleLogger(Level.Error))
```

Comment:
That does not seem right. What about the other things in `Emitter.symbolsRequirements`, such as `new java.lang.Class(Object)`?

Reply:
What about it? `Emitter.symbolRequirements` should only be symbols needed by the emitter (which we do not check here). The frontend should work with a different backend that has completely different symbol requirements.

Reply:
Looking at this more closely, I realize there are two kinds of requirements that the `Emitter` advertises:

* Requirements linked to the compilation strategy that the emitter uses, i.e., the `RuntimeLong` requirements
* Requirements linked to the semantics of certain IR constructs (everything else in `Emitter.symbolRequirements`)

The former can definitely be omitted in the `IRCheckerTest` and any other front-end-only tests.

The latter should *always* be required, and hence I believe they should be moved directly inside `Analyzer`. No matter what the compilation strategy of a back-end is, it *must* for example instantiate a `java.lang.Class` at some point (for `GetClass` and `ClassOf`).

So I guess the proper solution would be to add a first commit that moves all these other requirements into the `Analyzer`. Then the current contents of the PR are fine after that.

Reply:
Hmmm... I started looking at this and while I agree with this distinction, I believe the current code already meets it: None of the things the Emitter reaches are required in all cases and are merely implementation details of the `CoreJSLib`:

> instantiateClass(ObjectClass, NoArgConstructorName)
> classData(ObjectClass),

Needed for arrays which we do not DCE (but could).

> instantiateClass(CloneNotSupportedExceptionClass, NoArgConstructorName),

Needed for Object.clone() which we do not DCE (but could).

> All exceptions:

Implementation detail of how the Emitter actually throws them. If more stuff were implemented in IR, they would be reached "normally".

> instantiateClass(ClassClass, ObjectArgConstructorName)

Private contract between our implementation of java.lang.Class and the CoreJS lib of how class data is passed. The `ClassClass` itself could arguably be reached by the Analzyer itself *iff* any class data is accessed. But I'm not sure it is worth doing that.

I'm not quite sure how Hijacked classes change this picture, but I'm sure you are :)

Reply:
Hum ... I think you're write for `ObjectClass` and for `CloneNotSupportedException`. However,

> > All exceptions:
> 
> Implementation detail of how the Emitter actually throws them. If more stuff were implemented in IR, they would be reached "normally".

No, those exceptions are thrown as part of the semantics of certain IR Trees. For example `ArraySelect` can throw an `ArrayIndexOutOfBoundsException`. There's no way we could implement that "normally" in IR code, because `ArraySelect` cannot be broken into more primitive IR trees without breaking the IR semantics of arrays. In fact the `Analyzer` already does something smart for `ArithmeticException`, by recognizing `Int_/` and friends and reaching `ArithmeticException` from there. It could do so for `ArraySelect`s, but instead it chooses to always reach `ArrayIndexOutOfBoundsException` in order not to have to track them. (We did the tracking for `ArithmeticException` because they are not gated by undefined behavior, and would therefore always be required, but that does not change the fundamental analysis.)

> > instantiateClass(ClassClass, ObjectArgConstructorName)
> 
> Private contract between our implementation of java.lang.Class and the CoreJS lib of how class data is passed. The `ClassClass` itself could arguably be reached by the Analzyer itself _iff_ any class data is accessed. But I'm not sure it is worth doing that.

This is part of the public contract of the IR. It's even specified at https://lampwww.epfl.ch/~doeraene/thesis/sjsir-semantics/#sec-sjsir-classof

> I'm not quite sure how Hijacked classes change this picture, but I'm sure you are :)

Hijacked classes are always reached because JS code can instantiate them through just creating literal values ^^ As such they are already correctly reached by the `Analyzer`.